tm2py/components/network/transit/transit_assign.py

```diff
@@ -1,9 +1,632 @@
+        emmebank_path = self.get_abs_path(
+            self.controller.config.emme.transit_database_path
+        )
+        emmebank = self.controller.emme_manager.emmebank(emmebank_path)
+        use_ccr = False
+        if self.controller.iteration >= 1:
+            use_ccr = self.config.transit.use_ccr
+            self.demand.run()
+        else:
+            self.demand.create_zero_matrix(emmebank)
+        for self._time_period in self.time_period_names:
+            msg = f"Transit assignment for period {self._time_period}"
+            with self.logger.log_start_end(msg):
+                self._scenario = self.get_emme_scenario(emmebank, self._time_period)
+                if use_ccr:
+                    self._run_ccr_assign()
+                    self._calc_segment_ccr_penalties()
+                else:
+                    self._run_extended_assign()
+                if self.config.transit.output_stop_usage_path is not None:
+                    network, class_stop_attrs = self._calc_connector_flows()
+                    self._export_connector_flows(network, class_stop_attrs)
+                if self.config.transit.output_transit_boardings_path is not None:
+                    self._export_boardings_by_line()
+
+    @property
+    def _transit_classes(self) -> List[AssignmentClass]:
+        emme_manager = self.controller.emme_manager
+        if self.config.transit.use_fares:
+            fare_modes = _defaultdict(lambda: set([]))
+            network = self._scenario.get_partial_network(
+                ["TRANSIT_LINE"], include_attributes=False
+            )
+            emme_manager.copy_attribute_values(
+                self._scenario, network, {"TRANSIT_LINE": ["#src_mode"]}
+            )
+            for line in network.transit_lines():
+                fare_modes[line["#src_mode"]].add(line.mode.id)
+        else:
+            fare_modes = None
+        spec_dir = os.path.join(
+            os.path.dirname(self.config.emme.project_path), "Specifications"
+        )
+        transit_classes = []
+        for class_config in self.config.transit.classes:
+            transit_classes.append(
+                AssignmentClass(
+                    class_config,
+                    self.config.transit,
+                    self._time_period,
+                    self.controller.iteration,
+                    self._num_processors,
+                    fare_modes,
+                    spec_dir,
+                )
+            )
+        return transit_classes
+
+    @property
+    def _duration(self) -> float:
+        duration_lookup = dict(
+            (p.name, p.length_hours) for p in self.config.time_periods
+        )
+        return duration_lookup[self._time_period]
+
+    def _run_ccr_assign(self):
+        transit_classes = self._transit_classes
+        assign_transit = self.controller.emme_manager.tool(
+            "inro.emme.transit_assignment.capacitated_transit_assignment"
+        )
```

Comment:
Question for INRO: Is the only way to send a cost function as a string? This seems rather nonideal for many reasons.

Reply:
(ditto for headway cost function)